typings/styled-components.d.ts

```diff
@@ -24,10 +26,29 @@ export interface InterpolationFunction<P> {
   (props: P): Interpolation<P>;
 }
 
+type WithComponentOverloads<Tags, T> = {
+  [K in keyof Tags]: StyledComponentClass<Tags[K], T>;
+};
+
+type PropsToFunctions<P, T> = {
+  // [K in keyof P]: (props: ThemedOuterStyledProps<P, T>) => any;
+}
+
+export interface StyledComponentClass<P, T> extends ComponentClass<ThemedOuterStyledProps<P, T>> {
+  extend: ThemedStyledFunction<P, T>;
+
+  withComponent<K extends keyof HTMLTags>(tag: K): WithComponentOverloads<HTMLTags, T>[K];
+  withComponent<K extends keyof SVGTags>(tag: K): WithComponentOverloads<SVGTags, T>[K];
+  withComponent(element: ComponentClass<P>): StyledComponentClass<P, T>;
+  attrs(attrs: P | PropsToFunctions<ThemedOuterStyledProps<P, T>, T>): ThemedStyledFunction<P, T>;
```

Comment:
- `attrs` does not belong to the styled component class, it should be within `ThemedStyledFunction`
- the argument for `attrs` is an object that mixes: values for **declared** props, values for **fresh** props, and functions that accept existing inner props and return values for existing or fresh props
- `attrs` should return a themed styled function that carries information about provided and introduced props (styled function can create a styled component without specifying required props, and with using introduced props)
It's most likely impossible to define this properly in typescript type system. It currently lacks proper higher order functions support. I've played a bit with `attrs` definition yesterday and could only come up with the following:
```ts
type Attrs<P, A extends Partial<P>, T> = {
  [K in keyof A]: ((props: ThemedStyledProps<P, T>) => A[K]) | A[K];
};

export interface ThemedStyledFunction<P, T, O = P> {
  (strings: TemplateStringsArray, ...interpolations: Interpolation<ThemedStyledProps<P, T>>[]): StyledComponentClass<O, T>;
  <U>(strings: TemplateStringsArray, ...interpolations: Interpolation<ThemedStyledProps<P & U, T>>[]): StyledComponentClass<O & U, T>;
  attrs<U, A  extends Partial<P> = {}>(attrs: Attrs<P & U, A, T>): ThemedStyledFunction<P & A & U, T, O & U>;
}
```
- I had to introduce a new generic type parameter `O` as a type for _outer props_ of future style components
- I discovered that functions inside are not contextually typed, and filed a bug in TS https://github.com/Microsoft/TypeScript/issues/16209
- specifying `U` explicitly in `attrs` breaks inference for `A` because of https://github.com/Microsoft/TypeScript/issues/14400